cloudbuild.yaml

```diff
@@ -14,10 +14,118 @@
 
 
 steps:
-# Uses the docker build step to build an image called cloud-run-pubsub-service
-- name: 'gcr.io/cloud-builders/docker'
-  args: ['build', '-t', 'gcr.io/$PROJECT_ID/cloud-run-pubsub-service', './philips_hue_integration_example']
+- id: 'build docker image'
+  # Uses the docker build step to build an image
+  name: 'gcr.io/cloud-builders/docker'
+  args: ['build', '-t', 'gcr.io/$PROJECT_ID/${_IMAGE_NAME}', './philips_hue_integration_example']
 
-# cloud-run-pubsub-service is pushed to Container Registry
+- id: 'branch name'
+  name: 'alpine'
+  entrypoint: 'sh'  
+  args: 
+  - '-c'
+  - | 
+      echo "***********************"
+      echo "$BRANCH_NAME"
+      echo "***********************"
+
+- id: 'tf init'
+  name: 'hashicorp/terraform:0.12.28'
+  entrypoint: 'sh'
+  args: 
+  - '-c'
+  - |
+      if [ -d "environments/$BRANCH_NAME/" ]; then
+        cd environments/$BRANCH_NAME
+        terraform init -backend-config "bucket=$PROJECT_ID-tfstate"
+      else
+        for dir in environments/*/
+        do 
+          cd ${dir}   
+          env=${dir%*/}
+          env=${env#*/}
+          echo ""
+          echo "*************** TERRAFORM INIT ******************"
+          echo "******* At environment: ${env} ********"
+          echo "*************************************************"
+          terraform init -backend-config "bucket=$PROJECT_ID-tfstate" || exit 1
+          cd ../../
+        done
+      fi 
+
+# [START tf-refresh]
+- id: 'tf refresh'
+  name: 'hashicorp/terraform:0.12.28'
+  entrypoint: 'sh'
+  args: 
+  - '-c'
+  - |
+      if [ -d "environments/$BRANCH_NAME/" ]; then
+        cd environments/$BRANCH_NAME
+        terraform refresh -var="project=$PROJECT_ID"
+      else
+        for dir in environments/*/
+        do 
+          cd ${dir}   
+          env=${dir%*/}
+          env=${env#*/}
+          echo ""
+          echo "*************** TERRAFORM REFRESH ******************"
+          echo "******* At environment: ${env} ********"
+          echo "*************************************************"
+          terraform refresh -var="project=$PROJECT_ID" || exit 1
+          cd ../../
+        done
+      fi 
+# [END tf-refresh]
+
+# [START tf-plan]
+- id: 'tf plan'
+  name: 'hashicorp/terraform:0.12.28'
+  entrypoint: 'sh'
+  args: 
+  - '-c'
+  - | 
+      if [ -d "environments/$BRANCH_NAME/" ]; then
+        cd environments/$BRANCH_NAME
+        terraform plan -var="project=$PROJECT_ID"
+      else
```

Comment:
Are all of the else statements necessary? Maybe we can just document at the top of the file that the branch name must be a subdirectory in "environment/"?

Reply:
I mostly copied the format in this tutorial (https://cloud.google.com/solutions/managing-infrastructure-as-code#before-you-begin), which explains the reasoning:

"The reason terraform init and terraform plan run for all environments subfolders is to make sure that the changes being proposed hold for every single environment. This way, before merging the pull request, you can review the plans to make sure access is not being granted to an unauthorized entity, for example."

I think it might be a bit confusing so I can definitely add comments, but I believe it's trying to check if you're on the dev or prod branch, and if so it only does terraform init, plan, and apply for that branch. Otherwise, it runs init and plan for all subfolders in environments to make sure the changes hold for every environment. Does this seem to make sense? I can change the if statement so it explicitly checks that $BRANCH_NAME == 'dev'  or 'prod' if that makes it more clear.